metric_learn/rca.py

```diff
@@ -62,32 +79,42 @@ def fit(self, data, chunks):
         when ``chunks[i] == j``, point i belongs to chunklet j.
     """
     data, chunks, num_chunks, d = self._process_inputs(data, chunks)
-
-    # mean center
-    data -= data.mean(axis=0)
-
-    # mean center each chunklet separately
-    chunk_mask = chunks != -1
-    chunk_data = data[chunk_mask]
-    chunk_labels = chunks[chunk_mask]
-    for c in xrange(num_chunks):
-      mask = chunk_labels == c
-      chunk_data[mask] -= chunk_data[mask].mean(axis=0)
-
-    # "inner" covariance of chunk deviations
+    chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
     inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+    rank = np.linalg.matrix_rank(inner_cov)
+
+    # If the inner covariance matrix is not full rank,
+    # the input data are first projected with a PCA to a space of dimension rank.
+    M_pca = np.identity(d)
+    if rank < d:
+      pca = decomposition.PCA(n_components = rank)
+      pca.fit(data)
+      M_pca = pca.components_
+      data = pca.transform(data)
+      chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
+      inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+      rank = np.linalg.matrix_rank(inner_cov)
```

Comment:
I think taking 90% of the variance is too low, and telling the user to preprocess with PCA when we're already preprocessing with PCA seems strange. Maybe we could add a parameter `pca_comps=None` which would then be used for the PCA preprocessing:

```python
data = PCA(n_components=pca_comps).fit_transform(data)
```

Then, if the inner-cov is not full rank, we can tell the user to adjust `pca_comps`. We can also allow it to be dual-purpose (like sklearn's `n_components`), where a value between 0 and 1 indicates %variance, and an integer >= 1 indicates the number of components to use.

Other than that, I like your proposal.